File: java/DivideTwoIntegers.java
public class DivideTwoIntegers {
    public int divide(int dividend, int divisor) {
        boolean sign = true;
        long ldividend = (long) dividend;
        long ldivisor = (long) divisor;
        if (ldividend < 0) {
            ldividend = -ldividend;
            sign = !sign;
        }
        if (ldivisor < 0) {
            ldivisor = -ldivisor;
            sign = !sign;
        }
        if (ldivisor == 1) {
            if (sign) {
                return (int) ldividend;
            } else {
                return (int) -ldividend;
            }
        }
        long quotient = 0;
        long token = 1;
        while (ldivisor != 0) {
            if (ldividend >= ldivisor) {
                ldividend -= ldivisor;
                quotient += token;
            }
            if (ldividend > ldivisor) {
                ldivisor <<= 1;
                token <<= 1;
            } else if (ldividend < ldivisor) {
                ldivisor >>= 1;
                token >>= 1;
            }
        }
        if (sign) {
            return (int) quotient;
        } else {
            return (int) -quotient;
        }
    }

    public static void main(String[] args) {
        DivideTwoIntegers solution = new DivideTwoIntegers();
        System.out.println(Integer.MAX_VALUE);
        System.out.println(solution.divide(Integer.MAX_VALUE, -2));
        System.out.println(solution.divide(Integer.MIN_VALUE, 2));
    }
}

File: cc/divide_two_integers.cc
#include "base_include.h"

class DivideTwoIntegers {
public:
    int divide(int dividend, int divisor) {
        if (1 == divisor) {
            return dividend;
        }
        if (-1 == divisor) {
            return -dividend;
        }
        long long ldd = (long long) dividend;
        long long ldv = (long long) divisor;
        long long lq = 0;
        long long bin = 1;

        int sign = 1;
        if (ldd < 0) {
            sign = -sign;
            ldd = -ldd;
        }
        if (ldv < 0) {
            sign = -sign;
            ldv = -ldv;
        }
        while (ldd != 0) {
            if (ldd >= ldv) {
                ldd -= ldv;
                lq += bin;
                ldv <<= 1;
                bin <<= 1;
            } else {
                ldv >>= 1;
                bin >>= 1;
            }
        }
        if (-1 == sign) {
            return (int) -lq;
        } else {
            return (int) lq;
        }
    }
};

int main(int argc, char *argv[]) {
    return 0;
}
